File: backend/app/schemas/user_dashboard.py
from pydantic import BaseModel, Field


class UserDashboardChartItem(BaseModel):
    label: str
    value: int


class UserDashboardSummary(BaseModel):
    total_tasks: int = 0
    open_tasks: int = 0
    completed_tasks: int = 0
    total_requests: int = 0
    open_requests: int = 0
    completed_requests: int = 0
    sent_letters: int = 0
    received_letters: int = 0


class UserDashboardLetters(BaseModel):
    sent_by_type: list[UserDashboardChartItem] = Field(default_factory=list)
    received_by_type: list[UserDashboardChartItem] = Field(default_factory=list)
    sent_by_status: list[UserDashboardChartItem] = Field(default_factory=list)
    received_by_status: list[UserDashboardChartItem] = Field(default_factory=list)


class UserDashboardResponse(BaseModel):
    user_name: str
    summary: UserDashboardSummary
    task_statuses: list[UserDashboardChartItem] = Field(default_factory=list)
    request_statuses: list[UserDashboardChartItem] = Field(default_factory=list)
    top_requesters: list[UserDashboardChartItem] = Field(default_factory=list)
    top_recipients: list[UserDashboardChartItem] = Field(default_factory=list)
    requester_departments: list[UserDashboardChartItem] = Field(default_factory=list)
    request_departments: list[UserDashboardChartItem] = Field(default_factory=list)
    request_forms: list[UserDashboardChartItem] = Field(default_factory=list)
    monthly_tasks: list[UserDashboardChartItem] = Field(default_factory=list)
    monthly_requests: list[UserDashboardChartItem] = Field(default_factory=list)
    letters: UserDashboardLetters

File: frontend/src/features/userDashboard.ts
import client from "../api/client";
import { endpoints } from "../api/endpoints";
import { fetchLegacyUserDashboard } from "./userDashboardFallback";

export type DashboardChartItem = { label: string; value: number };

export type UserDashboardData = {
  user_name: string;
  summary: {
    total_tasks: number;
    open_tasks: number;
    completed_tasks: number;
    total_requests: number;
    open_requests: number;
    completed_requests: number;
    sent_letters: number;
    received_letters: number;
  };
  task_statuses: DashboardChartItem[];
  request_statuses: DashboardChartItem[];
  top_requesters: DashboardChartItem[];
  top_recipients: DashboardChartItem[];
  requester_departments: DashboardChartItem[];
  request_departments: DashboardChartItem[];
  request_forms: DashboardChartItem[];
  monthly_tasks: DashboardChartItem[];
  monthly_requests: DashboardChartItem[];
  letters: {
    sent_by_type: DashboardChartItem[];
    received_by_type: DashboardChartItem[];
    sent_by_status: DashboardChartItem[];
    received_by_status: DashboardChartItem[];
  };
};

export async function fetchUserDashboard(): Promise<UserDashboardData> {
  try {
    const { data } = await client.get<unknown>(endpoints.userDashboard);
    if (
      data &&
      typeof data === "object" &&
      "summary" in data &&
      "task_statuses" in data &&
      Array.isArray((data as Partial<UserDashboardData>).task_statuses)
    ) {
      return data as UserDashboardData;
    }
  } catch {
    // Older running backends do not expose the aggregated endpoint yet.
  }
  return fetchLegacyUserDashboard();
}
